dateutil/parser.py

```diff
@@ -742,318 +763,215 @@ def _parse(self, timestr, dayfirst=None, yearfirst=None, fuzzy=False,
                     # right.
-                    if i < len_l and l[i] in ('+', '-'):
-                        l[i] = ('+', '-')[l[i] == '+']
+                    if i+1 < len_l and l[i+1] in ('+', '-'):
+                        l[i+1] = ('+', '-')[l[i+1] == '+']
                         res.tzoffset = None
                         if info.utczone(res.tzname):
                             # With something like GMT+3, the timezone
                             # is *not* GMT.
                             res.tzname = None
 
-                    continue
 
                 # Check for a numbered timezone
-                if res.hour is not None and l[i] in ('+', '-'):
+                elif res.hour is not None and l[i] in ('+', '-'):
                     signal = (-1, 1)[l[i] == '+']
-                    i += 1
-                    len_li = len(l[i])
+                    len_li = len(l[i+1])
 
+                    # TODO: check that l[i+1] is integer?
                     if len_li == 4:
                         # -0300
-                        res.tzoffset = int(l[i][:2])*3600+int(l[i][2:])*60
-                    elif i+1 < len_l and l[i+1] == ':':
+                        hour_offset = int(l[i+1][:2])
+                        min_offset = int(l[i+1][2:])
+                    elif i+2 < len_l and l[i+2] == ':':
                         # -03:00
-                        res.tzoffset = int(l[i])*3600+int(l[i+2])*60
+                        hour_offset = int(l[i+1])
+                        min_offset = int(l[i+3]) # TODO: Check that l[i+3] is minute-like?
                         i += 2
                     elif len_li <= 2:
                         # -[0]3
-                        res.tzoffset = int(l[i][:2])*3600
+                        hour_offset = int(l[i+1][:2])
+                        min_offset = 0
                     else:
-                        return None, None
-                    i += 1
+                        raise InvalidDatetimeError(timestr)
 
-                    res.tzoffset *= signal
+                    res.tzoffset = signal * (hour_offset*3600 + min_offset*60)
 
+                    # TODO: Are we not requiring that res.tzname be None?
                     # Look for a timezone name between parenthesis
-                    if (i+3 < len_l and
-                        info.jump(l[i]) and l[i+1] == '(' and l[i+3] == ')' and
-                        3 <= len(l[i+2]) <= 5 and
-                        not [x for x in l[i+2]
-                             if x not in string.ascii_uppercase]):
+                    if (i+5 < len_l and
+                        info.jump(l[i+2]) and l[i+3] == '(' and l[i+5] == ')' and
+                        3 <= len(l[i+4]) <= 5 and
+                        all(x in string.ascii_uppercase for x in l[i+4])): # TODO: merge this with _could_be_tzname
                         # -0300 (BRST)
-                        res.tzname = l[i+2]
+                        res.tzname = l[i+4]
```

Comment:
These are because I am delaying incrementing `i` for as long as possible.  This decreases the total number of places where `i` is incremented at all (41 down from 51).  The closer `i` is to fixed within each iteration of the loop, the easier I find this to reason about.